storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3108,10 +3108,1655 @@ buf_block_t *fsp_get_latched_xdes_page(
+                                        &local_mtr, &err);
+  if (UNIV_UNLIKELY(!root))
+  {
+err_exit:
+    local_mtr.commit();
+    return err;
+  }
+
+  fseg_header_t *seg_header=
+    root->page.frame + PAGE_HEADER + PAGE_BTR_SEG_TOP;
+
+  buf_block_t *iblock;
+  fseg_inode_t *inode= fseg_inode_try_get(seg_header, 0, 0, &local_mtr,
+                                          &iblock, &err);
+  if (!inode)
+    goto err_exit;
+
+  uint16_t i_offset= uint16_t(inode - iblock->page.frame);
+
+  flst_validate(iblock, uint16_t(i_offset + FSEG_FREE), &local_mtr);
+  flst_validate(iblock, uint16_t(i_offset + FSEG_NOT_FULL), &local_mtr);
+  flst_validate(iblock, uint16_t(i_offset + FSEG_FULL), &local_mtr);
+
+  seg_header= root->page.frame + PAGE_HEADER + PAGE_BTR_SEG_LEAF;
+
+  inode= fseg_inode_try_get(seg_header, 0, 0, &local_mtr,
+                            &iblock, &err);
+  if (!inode)
+    goto err_exit;
+
+  i_offset= uint16_t(inode - iblock->page.frame);
+
+  flst_validate(iblock, uint16_t(i_offset + FSEG_FREE), &local_mtr);
+  flst_validate(iblock, uint16_t(i_offset + FSEG_NOT_FULL), &local_mtr);
+  flst_validate(iblock, uint16_t(i_offset + FSEG_FULL), &local_mtr);
+  local_mtr.commit();
+  return err;
+}
+#endif /* UNIV_DEBUG */
+
+void indexDefragmenter::store_child_pages(buf_block_t *block) noexcept
+{
+  rec_t *rec=
+    page_rec_get_next(page_get_infimum_rec(buf_block_get_frame(block)));
+  while (!page_rec_is_supremum(rec))
+  {
+    ulint len;
+    ulint offset=
+      rec_get_nth_field_offs_old(rec, rec_get_n_fields_old(rec) - 1,
+                                 &len);
+    ut_ad(len == 4);
+
+    ut_ad(offset < srv_page_size);
+    byte *field= rec + offset;
+    /* m_parent_pages[child_page_no] =
+       1st 32 bit to indicate offset in parent page
+       2nd 32 bit to indicate parent page number */
+    m_parent_pages[mach_read_from_4(field)]=
+      uint64_t(page_offset(field)) << 32 | block->page.id().page_no();
+    rec= page_rec_get_next(rec);
+  }
+}
+
+uint32_t indexDefragmenter::get_level_block(uint16_t level, dberr_t *err) noexcept
+{
+  mtr_t mtr;
+  mtr.start();
+  mtr_x_lock_index(m_index, &mtr);
+  uint32_t child_page_no= m_index->page;
+  while (1)
+  {
+    dberr_t err= DB_SUCCESS;
+    buf_block_t *block= buf_page_get_gen(page_id_t{0, child_page_no},
+                                         0, RW_S_LATCH, nullptr,
+                                         BUF_GET_POSSIBLY_FREED, &mtr,
+                                         &err);
+    if (!block)
+    {
+      child_page_no= FIL_NULL;
+      break;
+    }
+
+    page_t *page= buf_block_get_frame(block);
+    uint16_t cur_level= btr_page_get_level(page);
+    if (cur_level == level)
+      break;
+    rec_t *rec= page_rec_get_next(page_get_infimum_rec(page));
+    if (rec && !page_rec_is_supremum(rec))
+    {
+      ulint len;
+      ulint offset=
+        rec_get_nth_field_offs_old(rec, rec_get_n_fields_old(rec) - 1,
+                                   &len);
```

Comment:
Some of the code is assuming `ROW_FORMAT=REDUNDANT`, but some tries to handle all formats. Can you simplify this?
```c++
  const rec_t *rec= page_rec_get_next_low(page + PAGE_OLD_INFIMUM, false);
  if (rec && rec != page + PAGE_OLD_SUPREMUM)
```